ldap/servers/slapd/dyncerts.c

```diff
@@ -798,6 +797,786 @@ dyncerts_cert2entry(CERTCertificate *cert)
+        slapi_pblock_get(pb, SLAPI_ADD_ENTRY, &e) < 0 || (NULL == pdcerts)) {
+        rc = LDAP_OPERATIONS_ERROR;
+        goto done;
+    }
+    dn = slapi_sdn_get_dn(slapi_entry_get_sdn_const(e));
+
+    /* Check entry attributes */
+    (void) dyncert_nickname_from_dn(&n, sdn);
+    rc = dyncerts_check_entry(pb, e, &n, returntext, true);
+    if (rc) {
+        e = NULL; /* caller will free upon error */
+        goto done;
+    }
+    /* Check that entry does not exist */
+    if (dyncerts_find_entry(sdn, LDAP_SCOPE_BASE, NULL, &ss)) {
+        e = NULL; /* caller will free upon error */
+        rc = LDAP_ALREADY_EXISTS;
+        goto done;
+    }
+    rc = dyncerts_import_entry(e, returntext, false);
+done:
+    /* make sure OPRETURN and RESULT_CODE are set */
+    slapi_pblock_get(pb, SLAPI_PLUGIN_OPRETURN, &error);
+    if (rc) {
+        if (!error) {
+            slapi_pblock_set(pb, SLAPI_PLUGIN_OPRETURN, &rc);
+        }
+        slapi_log_err(SLAPI_LOG_DYC, "dyncerts_add",
+                      "Add operation of entry %s failed: rc=%d : %s\n",
+                      dn, rc, returntext);
+    }
+    ss_destroy(&ss);
+    free_config_info();
+    dyncert_nickname_free(&n);
+    slapi_send_ldap_result(pb, rc, NULL, returntext[0] ? returntext : NULL, 0, NULL);
+    /* The frontend does not free the added entry, so we should do it now */
+    if (e) {
+        slapi_entry_free(e);
+    }
+    slapi_pblock_set(pb, SLAPI_ADD_ENTRY, NULL);
+    return rc;
+}
+
+/* Create pseudo entry to add from modifiers */
+static int
+dyncerts_mods2entry(Slapi_PBlock *pb, Slapi_DN *sdn, LDAPMod **mods, Nickname_t *n, char *errmsg, Slapi_Entry **pte)
+{
+    const char *nickname = dyncert_nickname_from_dn(n, sdn);
+    char *dn = slapi_ch_strdup(slapi_sdn_get_dn(sdn));
+    Slapi_Entry *e = *pte = slapi_entry_alloc();
+    const char *val = NULL;
+    int rc = LDAP_SUCCESS;
+
+    slapi_entry_init(e, dn, NULL);
+    slapi_entry_add_string(e, "objectclass", "top");
+    slapi_entry_add_string(e, "objectclass", "extensibleobject");
+    if (!nickname) {
+        /* Should not happen (unless heap corruption) because we already
+         * tested that the modified entry exists
+         */
+        rc = LDAP_NAMING_VIOLATION;
+        goto done;
+    }
+    slapi_entry_add_string(e, DYCATTR_NICKNAME, nickname);
+    rc = slapi_entry_apply_mods(e, mods);
+    if (rc) {
+        goto done;
+    }
+    val = slapi_entry_attr_get_charptr(e, DYCATTR_NICKNAME);
+    if ((val == NULL) || (strcasecmp(val, nickname) != 0)) {
+        /* cn attribute was changed ! */
+        rc = LDAP_NAMING_VIOLATION;
+        goto done;
+    }
+    rc = dyncerts_check_entry(pb, e, n, errmsg, false);
+done:
+    return rc;
+}
+
+/* dyncerts_apply_cb callback to change trust flags */
+void
+dyncert_set_trust_cb(CertCtx_t *ctx)
+{
+    Slapi_Entry *e = ctx->arg;
+    CERTCertTrust trust = {0};
+    int rv = 0;
+
+    ctx->trust = slapi_entry_attr_get_charptr(e, DYCATTR_TRUST);
```

Comment:
Shouldn't free ctx->trust before setting it. Also if we return UNWILLING_TO_PERFORM, revert the original value.